File: 915.Partition-Array-into-Disjoint-Intervals.go
// https://leetcode.com/problems/partition-array-into-disjoint-intervals/
//
// algorithms
// Medium (40.45%)
// Total Accepted:    6,387
// Total Submissions: 15,790

package leetcode

func partitionDisjoint(A []int) int {
	length := len(A)

	bigest, smallest := make([]int, length), make([]int, length)

	bigest[0] = A[0]
	smallest[length-1] = A[length-1]

	for i := 1; i < length; i++ {
		if A[i] > bigest[i-1] {
			bigest[i] = A[i]
		} else {
			bigest[i] = bigest[i-1]
		}
	}

	for i := length - 2; i >= 0; i-- {
		if A[i] < smallest[i+1] {
			smallest[i] = A[i]
		} else {
			smallest[i] = smallest[i+1]
		}
	}

	for i := 0; i < length-1; i++ {
		if bigest[i] <= smallest[i+1] {
			return i + 1
		}
	}

	return 0
}

File: 915.Partition-Array-into-Disjoint-Intervals.py
# https://leetcode.com/problems/partition-array-into-disjoint-intervals/
#
# algorithms
# Medium (43.82%)
# Total Accepted:    13,024
# Total Submissions: 29,719


class Solution(object):
    def partitionDisjoint(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        length = len(A)
        idx = 0
        tmp = 0
        cur_max = A[0]

        while idx < length:
            cur_max_tmp = []
            for i in xrange(idx + 1, length):
                if A[i] < cur_max:
                    tmp = i

                elif A[i] > cur_max:
                    cur_max_tmp += i,

            if not cur_max_tmp or cur_max_tmp[0] > tmp:
                break

            idx = tmp
            for i in cur_max_tmp:
                if i > tmp:
                    break

                cur_max = max(cur_max, A[i])

        return tmp + 1
